core/src/main/java/com/linecorp/armeria/client/DefaultEventLoopScheduler.java

```diff
@@ -0,0 +1,388 @@
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Streams;
+
+import com.linecorp.armeria.common.util.ReleasableHolder;
+
+import io.netty.channel.EventLoop;
+import io.netty.channel.EventLoopGroup;
+
+final class DefaultEventLoopScheduler implements EventLoopScheduler {
+
+    private static final long CLEANUP_INTERVAL_NANOS = Duration.ofMinutes(1).toNanos();
+
+    private final List<EventLoop> eventLoops;
+
+    private final int maxEventLoopPerEndpoint;
+
+    private final Map<String, Integer> maxEventLoops;
+
+    private final Map<String, State> states = new ConcurrentHashMap<>();
+
+    private final AtomicInteger currentIndex = new AtomicInteger();
+
+    private int counter;
+    private volatile long lastCleanupTimeNanos = System.nanoTime();
+
+    DefaultEventLoopScheduler(EventLoopGroup eventLoopGroup, int maxEventLoopPerEndpoint,
+                              Map<String, Integer> maxEventLoops) {
+        eventLoops = Streams.stream(eventLoopGroup)
+                            .map(EventLoop.class::cast)
+                            .collect(toImmutableList());
+        checkArgument(maxEventLoopPerEndpoint <= eventLoops.size(),
+                      "maxEventLoopPerEndpoint: %s (expected: <= %s)", maxEventLoopPerEndpoint,
+                      eventLoops.size());
+        if (maxEventLoopPerEndpoint == 0) {
+            this.maxEventLoopPerEndpoint = eventLoops.size();
+        } else {
+            this.maxEventLoopPerEndpoint = maxEventLoopPerEndpoint;
+        }
+
+        for (Map.Entry<String, Integer> entry : maxEventLoops.entrySet()) {
+            checkArgument(entry.getValue() <= eventLoops.size(),
+                          "maxEventLoopPerEndpoint for %s: %s (expected: <= %s)",
+                          entry.getKey(), entry.getValue(), eventLoops.size());
+        }
+        this.maxEventLoops = ImmutableMap.copyOf(maxEventLoops);
+    }
+
+    @Override
+    public ReleasableHolder<EventLoop> acquire(Endpoint endpoint) {
+        requireNonNull(endpoint, "endpoint");
+        final State state = state(endpoint);
+        final Entry acquired = state.acquire();
+        cleanup();
+        return acquired;
+    }
+
+    @VisibleForTesting
+    List<Entry> entries(Endpoint endpoint) {
+        return state(endpoint).entries();
+    }
+
+    private State state(Endpoint endpoint) {
+        final String authority = endpoint.authority();
+        final State state = states.get(authority);
+        if (state != null) {
+            return state;
+        }
+
+        final Integer maxEventLoop = maxEventLoops.get(authority);
```

Comment:
We have to take the default port into account so that `foo.com:80` and `foo.com` are not treated as different endpoints. To fix this:

- Add `SessionProtocol` as an input parameter to `EventLoopScheduler.acquire()` and `ClientFactory.acquireEventLoop()`.
- Handle the cases where:
  - the `Endpoint` or authority string specified in `ClientFactoryBuilder.maxNumEventLoopsPerEndpoint()` has/doesn't have a port number. (If no port number is specified, we should ignore the port number in the `Endpoint` given to `acquire()`.
  - If the `Endpoint` given to `acquire()` does not have a port number, we need to get one from the given `SessionProtocol`.

Reply:
Added `SessionProtocol` to `EventLoopScheduler.acquire()` and change the key of the `states` map from `authority` to `authority` with `isTLS`.